### DISCUSSION TITLE
Global Timer Somewhere?

### DISCUSSION THREAD

**Discussion by MPatriot** [oqtane/oqtane.framework#1919] [Q&A]
Hello all,

I was wondering if there is a timer running somewhere that can be subscribed to?

For instance, how does a job / HostedServiceBase know when to check how much time has elapsed?  

Or if I just wanted a clock showing with H/M/S ticking?  Sure I could put a timer on a razor page but that would get awful expensive if multiple timers were running.

Just wondering,

Thanks,

-Mark

**Comment by leigh-pointer** [comment]
Hi @MPatriot look in the code location ..\oqtane.framework\Oqtane.Server\Infrastructure\Jobs\NotificationJob.cs

Jobs are quite easy to set up.

**Reply by MPatriot** [reply]
Thanks Leigh,

I have a job setup in a module but if I wanted to subscribe to whatever is telling it to fire (minimum 1m) to check what time it is now, where would I find that?  I see that NotificationJob uses HostedServiceBase which sets all that, including job.NextExecution but it seems some event has to be firing all that

-Mark

**Reply by leigh-pointer** [reply]
@MPatriot you need to create a job.  Create a new class in your module server project and inherit from `HostedServiceBase` from there you should follow the framework class  ..\oqtane.framework\Oqtane.Server\Infrastructure\Jobs\NotificationJob.cs

**Reply by MPatriot** [reply]
Yeah...
I get all that but let me try to say this again. If I wanted a clock with hours, minutes and seconds showing on a page it doesn't seem like a job is the right thing for that, ultimately jobs get saved in the database through the job repository and there is a minimum of 1m, that would make for a heck of a lot every second (even if I modified all of that) what I saying is there has to be something firing, like aTimer.Elapsed += OnTimedEvent; that checks every job to see if it is time to fire, though its increment might be more than a 1000ms so it might not be usable and a separate timer type thing might be better.  It would definitely be better if it was a client side thing but right now, I am doing everything server based and didn't want to add more of a load there.

Thanks

**Reply by MPatriot** [reply]
Here is one more example, right in what we are talking about...

The Scheduled Jobs page has a Next Execution column on it, which does not update after the job executes without hitting the refresh button.  I am sure I could update the page after job executes and make it more dynamic but this kind of thing could be throughout any app I develop and so I was really just asking if there was already a global timer event firing or do I need to add one ?

**Reply by leigh-pointer** [reply]
@MPatriot there is no single timer running the Scheduler creates a Task for each job. check out the code here ..\oqtane.framework\Oqtane.Server\Infrastructure\Jobs\HostedServiceBase.cs

If you require a timer you will have to create one. AFAICS

**Reply by sbwalker** [reply]
@MPatriot it sounds what you are looking for is real-time communication between the server and client (as they are in fact independent workloads). There are a number of ways to accomplish this but nothing that is included in Oqtane. Options would include SignalR, polling, event bus, etc...

In regards to scheduled jobs there is a timer of sorts which runs constantly in HostedServiceBase - waking up every minute to determine if a job needs to execute (code below). The NextExecution is only updated after the current job execution is completed. And the Scheduled Jobs user interface is not automatically updated when a job completes because the job is running asynchronously on the server and there is no real-time communication channel between the client and asynchronous server process. 

```
                 while (!stoppingToken.IsCancellationRequested)
                {
                    ...
                    // wait 1 minute
                    await Task.Delay(TimeSpan.FromMinutes(1), stoppingToken);
                }
```

**Reply by MPatriot** [reply]
Thanks Shaun,
Yes, I found system.threading and system.threading.tasks used many places, like when checking if a file is locked ect...
For a "Clock"  I might set up a timer in a singleton service and then any any modules that wanted to subscribe to it could as long as I disposed of that subscription everywhere it is used to prevent memory leaks.  This is probably preferable than multiple timers per how many ever users are accessing a page / module, i have done this in other Blazor projects and it seems to have been fine

I really like what you have going on here, have done similar things with MVC projects and loading different master pages per domain names and different tenants as you call them with different connection strings but you really have a nice and well rounded back end and UI set of building blocks to work from.

Thanks!

-Mark

**Comment by W6HBR** [comment]
@MPatriot 
Did you ever implement your singleton "Clock" timer?  If so, I'm curious about the details of how you went about doing that.

I have created a timer in the OnInitializedAsync task and it seems to function quite well.  I am concerned though with whether it would scale well if I have more than a handful of clients.

`       
 // Use Async timer to update the running clock on the screen.
// NOTE: MUST CALL StateHasChanged() to update display because this is triggered by a Timer instead of a user event or UI element.

        ClockTimer = new System.Threading.Timer(async (object stateInfo) =>
        {
            _Header_Date = DateTime.Now.ToString("yyyy-MM-dd");
            _Header_Time = DateTime.Now.TimeOfDay.ToString("hh\\:mm\\:ss");
            await InvokeAsync(StateHasChanged);
        },  new System.Threading.AutoResetEvent(false), 1000, 1000); // fire every 1000 milliseconds

`

**Reply by sbwalker** [reply]
@W6HBR you are creating a new timer instance each time this component is loaded... and you are never disposing it. This approach will result in a memory leak.

**Reply by sbwalker** [reply]
This implementation avoids the memory leak:

```
@implements IDisposable
 
@code {
    System.Threading.Timer timer;
    int timeout = 2000; // 2 seconds
 
    protected override void OnInitialized()
    {
        timer = new System.Threading.Timer(async _ =>
        {
            // your logic

        }, null, timeout, timeout);
    }
 
    public void Dispose()
    {
        timer?.Dispose();
    }
}
```

**Comment by W6HBR** [comment]
@sbwalker  Excellent! Thank you.  That was a concern I had with this code.
